test/conformance/certificate/util.go

```diff
@@ -0,0 +1,146 @@
+	name := test.ObjectNameForTest(t)
+
+	cert := &v1alpha1.Certificate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: test.ServingNamespace,
+			Annotations: map[string]string{
+				networking.CertificateClassAnnotationKey: getCertClass(t, clients),
+			},
+		},
+		Spec: v1alpha1.CertificateSpec{
+			DNSNames:   dnsNames,
+			SecretName: name,
+		},
+	}
+
+	test.CleanupOnInterrupt(func() {
+		clients.NetworkingClient.Certificates.Delete(cert.Name, &metav1.DeleteOptions{})
+		clients.KubeClient.Kube.CoreV1().Secrets(test.ServingNamespace).Delete(cert.Spec.SecretName, &metav1.DeleteOptions{})
+	})
+	cert, err := clients.NetworkingClient.Certificates.Create(cert)
+	if err != nil {
+		t.Fatalf("Error creating Certificate: %v", err)
+	}
+
+	return cert, func() {
+		err := clients.NetworkingClient.Certificates.Delete(cert.Name, &metav1.DeleteOptions{})
+		if err != nil {
+			t.Errorf("Error cleaning up Certificate %s: %v", cert.Name, err)
+		}
+		secretName := cert.Spec.SecretName
+		err = clients.KubeClient.Kube.CoreV1().Secrets(test.ServingNamespace).Delete(secretName, &metav1.DeleteOptions{})
+		if err != nil {
+			t.Errorf("Error cleaning up Secret %s for Certificate %s: %v", secretName, cert.Name, err)
+		}
+
+	}
+}
+
+// WaitForCertificateState polls the status of the Certificate called name from client
+// every PollInterval until inState returns `true` indicating it is done, returns an
+// error or PollTimeout. desc will be used to name the metric that is emitted to
+// track how long it took for name to get into the state checked by inState.
+func waitForCertificateState(client *test.NetworkingClients, name string, inState func(r *v1alpha1.Certificate) (bool, error), desc string) error {
+	span := logging.GetEmitableSpan(context.Background(), fmt.Sprintf("WaitForCertificateState/%s/%s", name, desc))
+	defer span.End()
+
+	var lastState *v1alpha1.Certificate
+	waitErr := wait.PollImmediate(test.PollInterval, test.PollTimeout, func() (bool, error) {
+		var err error
+		lastState, err = client.Certificates.Get(name, metav1.GetOptions{})
+		if err != nil {
+			return true, err
+		}
+		return inState(lastState)
+	})
+
+	if waitErr != nil {
+		return fmt.Errorf("certificate %q is not in desired state, got: %+v: %w", name, lastState, waitErr)
+	}
+	return nil
+}
+
+func waitForSecret(client *test.Clients, name string, desc string) error {
+	span := logging.GetEmitableSpan(context.Background(), fmt.Sprintf("waitForSecret/%s/%s", name, desc))
+	defer span.End()
+
+	waitErr := wait.PollImmediate(test.PollInterval, test.PollTimeout, func() (bool, error) {
+		_, err := client.KubeClient.Kube.CoreV1().Secrets(test.ServingNamespace).Get(name, metav1.GetOptions{})
+
+		if apierrs.IsNotFound(err) {
+			return false, nil
+		}
+
+		return true, err
+	})
+
+	if waitErr != nil {
+		return fmt.Errorf("secret %q not found: %w", name, waitErr)
+	}
+	return nil
+}
+
+func certHasHTTP01Challenges(cert *v1alpha1.Certificate) (bool, error) {
```

Comment:
I'd inline this at the call site.

Reply:
Do you mean as part of the call to `waitForCertificateState` or do you mean define it inside of `TestHTTP01Challenge`?

Reply:
Former. Right in the call site.

Reply:
Done